### PR TITLE
fixs for Refinery spawned harvesters should prioritize lower ore 

### LINKED ISSUE #14827: Refinery spawned harvesters should prioritize lower ore
Upon spawning, harvesters will move around the entire refinery if there is ore above and to the left. This makes ore trucks move past closer ore.

As most maps have starting ore to the left/right of players, the players with ore on the left will get behind if they don't immediatley micro their harvestor upon building a refinery. 

trucks

### PR DESCRIPTION
**Pull request by CJure**
Closes #14827 Refinery spawned harvesters should prioritize lower

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,7 +11,7 @@ obj
 *.CodeAnalysisLog.xml
 *.lastcodeanalysissucceeded
 _ReSharper.*/
-/.vs
+/.vs
```
Comment (GraionDilach) [inline .gitignore]: Revert this one.

File: .gitignore [hunk 2]
```diff
@@ -82,6 +82,7 @@ StyleCopViolations.xml
 
 # Support directory
 /Support
-
-# IntelliJ files
-.idea
+
```
Comment (GraionDilach) [inline .gitignore]: And these as well.

**Comment by GraionDilach** [comment]
>OpenRA.Mods.Common/Activities/FindResources.cs:L169: [SA1513] Statements or elements wrapped in curly brackets must be followed by a blank line.
OpenRA.Mods.Common/Activities/FindResources.cs:L173: [SA1508] A closing curly bracket must not be preceded by a blank line.

**Comment by fruestueck** [comment]
Welcome CJure,
you can use `make.cmd` `check` before uploading a commit to find and avoid style violations.
I'd also avoid #[issueNr] in a PR-title and commit-description.

**Comment by CJure** [comment]
@GraionDilach fixed
@fruestueck tnx for make.cmd, didn't know about that. Removd PR from title.

**Review by Smittytron — APPROVED** [review]
Tested in game. Works as advertised, harvesters no longer run all around the refinery upon spawning.

File: OpenRA.Mods.Common/Activities/FindResources.cs [hunk 3]
```diff
@@ -152,5 +152,16 @@ public override IEnumerable<Target> GetTargets(Actor self)
...
 			yield return Target.FromCell(self.World, self.Location);
 		}
+
+		public CPos GetFromSearchLocation(Actor self)
+		{
+			if (harv.LastLinkedProc != null)
+				return harv.LastLinkedProc.Location + harv.LastLinkedProc.Trait<IAcceptResources>().DeliveryOffset;
+			else if (harv.LinkedProc != null)
+				return harv.LinkedProc.Location + harv.LinkedProc.Trait<IAcceptResources>().DeliveryOffset;
+			else if (harv.LastOrderLocation != null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/FindResources.cs]: Why was this moved to lowest priority?  Won't this break all the cases where we *don't* want the closest ore?

Comment (reaperrr) [reply OpenRA.Mods.Common/Activities/FindResources.cs]: I agree. Using DeliveryOffset instead of Location should be the only change here.

LastOrderLocation needs to remain on top, else we'll definitely run into regressions.

**Comment by CJure** [comment]
Changed back

File: OpenRA.Mods.Common/Activities/FindResources.cs [hunk 4]
```diff
@@ -152,5 +152,16 @@ public override IEnumerable<Target> GetTargets(Actor self)
 		{
 			yield return Target.FromCell(self.World, self.Location);
 		}
+
+		public CPos GetFromSearchLocation(Actor self)
```
Comment (Unit158) [inline OpenRA.Mods.Common/Activities/FindResources.cs]: This should be private.

File: OpenRA.Mods.Common/Activities/FindResources.cs [hunk 5]
```diff
@@ -152,5 +152,16 @@ public override IEnumerable<Target> GetTargets(Actor self)
...
+		{
+			if (harv.LastOrderLocation != null)
```
Comment (Unit158) [inline OpenRA.Mods.Common/Activities/FindResources.cs]: Given that this is a nullable type, I'd check if it has a value

File: OpenRA.Mods.Common/Activities/FindResources.cs [hunk 6]
```diff
@@ -152,5 +152,16 @@ public override IEnumerable<Target> GetTargets(Actor self)
...
+				return (CPos)harv.LastOrderLocation;
```
Comment (Unit158) [inline OpenRA.Mods.Common/Activities/FindResources.cs]: and get the value here instead of casting

**Review by reaperrr — DISMISSED** [review]

File: OpenRA.Mods.Common/Activities/FindResources.cs [hunk 7]
```diff
@@ -152,5 +152,16 @@ public override IEnumerable<Target> GetTargets(Actor self)
 		{
 			yield return Target.FromCell(self.World, self.Location);
 		}
+
+		private CPos GetFromSearchLocation(Actor self)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/FindResources.cs]: No explicit `private` keyword, please (that is the default already). Can we also rename this to `GetSearchFromLocation`? It makes more sense imho and the local variable is also called `searchFromLoc`.

**Review by abcdefg30 — DISMISSED** [review]
After the comment is addressed.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/602e97f7326dbee2b96c584c0c3919632b847af7...a6d4a0c4ae783945f64d2261cfb809a74fc9ed42)
